### PR TITLE
Dev FaceSeg

### PR DESCRIPTION
**Pull request by JiaojiaoYe1994**
# FaceSeg 算法提交

开发，提交文件checklist
* [x] 训练脚本
* [x] 预训练模型
* [x] 预测脚本
* [x] requirements.txt
* [x] README

迁移进度和问题：

－　目前推理部分已完成，之后可补充一个视频人脸分割以及实时人脸分割的demo。
－　训练部分正在迁移，遇到loss.require_grad是False的问题，反向反向图没正确构建，目前正在排查原因。

### CODE REVIEW

File: FaceSeg/utils/plot.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# -*- coding:utf-8 -*-
```
Comment (Ldpe2G) [inline FaceSeg/utils/plot.py:1]: 这个文件可以删除了

File: FaceSeg/models/pytorch_resnet50.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
+# -*- coding:utf-8 -*-
```
Comment (Ldpe2G) [inline FaceSeg/models/pytorch_resnet50.py:1]: 这个文件可以删除了

File: FaceSeg/utils/plot.py (hunk 1 above)
Comment (JiaojiaoYe1994) [reply FaceSeg/utils/plot.py:1]: 好的

File: FaceSeg/models/pytorch_resnet50.py (hunk 2 above)
Comment (JiaojiaoYe1994) [reply FaceSeg/models/pytorch_resnet50.py:1]: 我删一下

File: FaceSeg/README.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+bash infer.sh
+```
+
+
+## Train 
+
+### Data preparation
+
+We combine several dataset and preprocess them, for training and evaluating our 
+face segmentation model. We provide the final compressed dataset, please download dataset from [here](https://oneflow-static.oss-cn-beijing.aliyuncs.com/train_data_zjlab/faceseg_d~~~~ata.zip) , unzip it and put it in `./`
```
Comment (Ldpe2G) [inline FaceSeg/README.md:41]: 数据集路径是不是有问题

File: FaceSeg/infer.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+
+# model_path = "linknet.pth"
+def main(args):
+    flow.env.init()
+    flow.enable_eager_execution()
+
+    # set path=0 for webcam or set to a video file
+    source = "result_video.avi"
+    source = '/home/jiaojiao/jiaojiao/project/video_source/source1.avi'
```
Comment (Ldpe2G) [inline FaceSeg/infer.py:107]: 不要有这种写死的路径，测试视频也要上传到阿里云，不然用户跑不了你这个脚本

File: FaceSeg/README.md (hunk 3 above)
Comment (JiaojiaoYe1994) [reply FaceSeg/README.md:41]: 已去除。

File: FaceSeg/infer.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
+# -*- coding:utf-8 -*-
+import sys
+
+import cv2
+import time
+import numpy as np
+from PIL import Image
+import argparse
+
+import oneflow.experimental as flow
```
Comment (Ldpe2G) [inline FaceSeg/infer.py:10]: experimental 已经去掉了, 可以更新下oneflow

File: FaceSeg/infer.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                break
+            self.frames_count += 1
+        self.terminate(camera)
+
+    def terminate(self, camera):
+        cv2.destroyAllWindows()
+        camera.release()
+
+
+# model_path = "linknet.pth"
```
Comment (Ldpe2G) [inline FaceSeg/infer.py:100]: 无用的注释代码删掉

File: FaceSeg/models/LinkNet34.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    def __init__(self, num_classes=1, num_channels=3, pretrained=False, pretrained_model_path = ""):
+        super().__init__()
+        assert num_channels == 3
+        self.num_classes = num_classes
+        filters = [64, 128, 256, 512]
+        resnet = resnet50()
+
+        # if pretrained=True, load pretrained resnet model
+        if pretrained:
+            # resnet = load_model_params(resnet,pretrained_model_path)
```
Comment (Ldpe2G) [inline FaceSeg/models/LinkNet34.py:68]: 删除

File: FaceSeg/models/resnet50.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+    # print('layers is: '+str(layers))
+    model = ResNet(block, layers, **kwargs)
+    return model
+
+# def resnet50(**kwargs: Any) -> ResNet:
+#     r"""ResNet-5
+#     `"Deep Residual Learning for Image Recognition" <https://arxiv.org/pdf/1512.03385.pdf>`_.
+#     """
+#     return _resnet('resnet50', Bottleneck, [3, 4, 6, 3], **kwargs)
+
```
Comment (Ldpe2G) [inline FaceSeg/models/resnet50.py:238]: 以上删

File: FaceSeg/models/resnet50.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        return x
+
+def _resnet(
+    arch: str,
+    block: Type[Union[BasicBlock, Bottleneck]],
+    layers: List[int],
+    **kwargs: Any
+) -> ResNet:
+    # print('layers is: '+str(layers))
```
Comment (Ldpe2G) [inline FaceSeg/models/resnet50.py:229]: 删

File: FaceSeg/train.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+            of_sgd.step()
+            cosine.step()
+            end_t = time.time()
+            l = loss.numpy()[0]
+            of_losses.append(l)
+            sys.stdout.write(f'\rEpoch: {epoch} ---- Loss: {round(epoch_loss / (b + 1), 4)} ----- num: {b}')
+            sys.stdout.flush()
+
+        print("epoch %d done, start validation" % epoch)
+    # flow.save(linknet34_module.state_dict(), "linknet34_oneflow_training_model_50_test3")
```
Comment (Ldpe2G) [inline FaceSeg/train.py:110]: 删

File: FaceSeg/train.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+    writer = open("of_losses.txt", "w")
+    for o in of_losses:
+        writer.write("%f\n" % o)
+    writer.close()
+
+if __name__ == "__main__":
+    args = _parse_args()
+    main(args)
+
```
Comment (Ldpe2G) [inline FaceSeg/train.py:121]: 下面多余空行

File: FaceSeg/utils/load_model_params.py [hunk 12]
```diff
@@ -0,0 +1,25 @@
+# -*- coding:utf-8 -*-
+import torch
+import time
+
```
Comment (Ldpe2G) [inline FaceSeg/utils/load_model_params.py:4]: 这个脚本没用了好像

File: FaceSeg/infer.py (hunk 4 above)
Comment (JiaojiaoYe1994) [reply FaceSeg/infer.py:107]: 已上传测试视频啦

File: FaceSeg/infer.py (hunk 6 above)
Comment (JiaojiaoYe1994) [reply FaceSeg/infer.py:100]: 好的

File: FaceSeg/train.py (hunk 11 above)
Comment (JiaojiaoYe1994) [reply FaceSeg/train.py:121]: 已删

File: FaceSeg/utils/load_model_params.py (hunk 12 above)
Comment (JiaojiaoYe1994) [reply FaceSeg/utils/load_model_params.py:4]: 是的 已删

File: FaceSeg/infer.py (hunk 5 above)
Comment (JiaojiaoYe1994) [reply FaceSeg/infer.py:10]: 需要更新到哪个版本？

File: FaceSeg/infer.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+        self.terminate(camera)
+
+    def terminate(self, camera):
+        cv2.destroyAllWindows()
+        camera.release()
+
+
+def main(args):
+    # flow.env.init()
+    # flow.enable_eager_execution()
```
Comment (Ldpe2G) [inline FaceSeg/infer.py:103]: 这里忘记删了

Comment (JiaojiaoYe1994) [reply FaceSeg/infer.py]: 好了

**Review by Ldpe2G — APPROVED** [review]
